### ISSUE TITLE
anything is assignable to T[P] where T extends null/undefined/void/object/never

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#15371]
### case 1
```typescript
export function test<T extends object, P extends keyof T>(
    whatever: T, name: P
): T[P] {
    const result = Math.random() > 0.5 ? whatever[name] : 'Hello!'; // <-- expected T[P] | string, actual T[P]
    return result; // <-- returning unsound result
}
```

### case 2
```typescript
export function oneOrAnother<A, B>(one: A, another: B): A | B {
    return Math.random() > 0.5 ? one : another;
}

export function test<T extends object, P extends keyof T>(
    whatever: T, name: P
): T[P] {
    const result = oneOrAnother(whatever[name], 'Hello!'); // <-- result is clearly string | T[P] as expected
    return result; // <-- expected a type error, actual returning string | T[P] instead of T[P] no problem
}
```

**Comment by MrAndersen1** [comment]
You probably already found this out but these issues arise from extending object.
What one gains from extending object and why these issues arise when you do it is beyond me however.

**Comment by sandersn** [comment]
Here's a smaller repro:

```ts
function test<T extends object, P extends keyof T>(x: T[P] | string, y: T[P]): void {
    y = x; // should error
}
```

**Comment by sandersn** [comment]
To check assignability of `string` to `T[P]`, the compiler checks whether `string` is assignable to the constraint of `T[P]`. But `getConstraintOfIndexedAccess` results in `anyType`: 

* the type constraint of `T[P]` is `object[string]`
* But `object` doesn't have a string index signature.
* so `getPropertyTypeForIndexType` returns `anyType`.

So `string` is assignable to `T[P]` because the apparent type of T[P] is `any`, which is incorrect. I think `getConstraintOfIndexedAccess` is incorrect but I'm not yet sure what the fix is.

**Comment by sandersn** [comment]
Fix is up at #17870.

**Comment by sandersn** [comment]
The previous PR got merged by mistake and is now reverted; #17912 is the PR with the fix now.

**Comment by aleksey-bykov** [comment]
looks like this pr might be the reason to https://github.com/Microsoft/TypeScript/issues/12215#issuecomment-357051029
